Introduce information about DslMarker with function types (#4679)
JetBrains/kotlin-web-site: docs/topics/type-safe-builders.md

}
```

You can also apply the `@DslMarker` annotation directly to [function types](lambdas.md#function-types).
To do so, annotate a `@DslMarker` annotation with `@Target(AnnotationTarget.TYPE)`:

```kotlin
@Target(AnnotationTarget.TYPE)
@DslMarker
annotation class HtmlTagMarker
```

> This allows the `@DslMarker` annotation to be applied to function types, most commonly to lambdas with receivers. For example:

Reviewer: ```suggestion
As a result, the `@DslMarker` annotation can be applied to function types, most commonly to lambdas with receivers. For example:
```

```kotlin
fun html(init: @HtmlTagMarker HTML.() -> Unit): HTML { ... }

fun HTML.head(init: @HtmlTagMarker Head.() -> Unit): Head { ... }

fun Head.title(init: @HtmlTagMarker Title.() -> Unit): Title { ... }
```

When you call these functions, the `@DslMarker` annotation restricts access to outer receivers in the body of a lambda marked with it unless you specify them explicitly:

```kotlin
html {
    head {
        title {
            // Access to title, head or other functions of outer receivers is restricted here.
        }
    }
}
```

> This ensures that only the nearest receiver’s members and extensions are accessible within a lambda, preventing unintended interactions between nested scopes.

Reviewer: ```suggestion
Only the nearest receiver's members and extensions are accessible within a lambda, preventing unintended interactions between nested scopes.
```
Removing "this ensures" because it's not 100% clear what "this" refers to.

Author: good point - updated it

}
```

You can also apply the `@DslMarker` annotation directly to [function types](lambdas.md#function-types).
To do so, annotate a `@DslMarker` annotation with `@Target(AnnotationTarget.TYPE)`:

```kotlin
@Target(AnnotationTarget.TYPE)
@DslMarker
annotation class HtmlTagMarker
```

This allows the `@DslMarker` annotation to be applied to function types, most commonly to lambdas with receivers. For example:

```kotlin

> fun HTML.head(init: @HtmlTagMarker Head.() -> Unit): Head {

Reviewer 2: NB: I do not know what is the preferred way of doing code examples in our technical documentation, you are the expert here and your opinion is the most important =)

I would prefer to have a more complete example w.r.t. `@DslMarker`, meaning we should also define the `html` function

```
fun html(init: @HtmlTagMarker HTML.() -> Unit): HTML {
    val html = HTML()
    html.init()
    return html
}
```

Reviewer 2: Also, maybe we could keep the bodies of the functions empty or as `TODO()`? Their definitions are not that important w.r.t. DSL scoping rules, and the ones we have now are kinda not entirely correct.

    val head = Head()
    head.init()
    return head
}

fun Head.title(init: @HtmlTagMarker Title.() -> Unit): Title {
    val title = Title()
    title.init()
    return title
}
```

> When you call these functions, the `@DslMarker` annotation restricts access to outer receivers unless you specify them explicitly:

Reviewer 2: Maybe explain this a bit, as in "...annotation restricts access to outer receivers [within the body of a `@DslMarker`-marked lambdas] unless you...?

```kotlin
html {
    head {
        title {

>             // Access to head or outer html members is restricted here.

Reviewer 2: ```suggestion
            // Access to `title`, `head` or other functions on outer receivers is restricted here.
```

        }
    }
}
```

> This ensures that only the nearest receiver’s members are accessible within a lambda, preventing unintended interactions between nested scopes.

Reviewer 2: It's not only about members, it's about all available callables, meaning extensions are also included in the restrictions.

Author: Great point - thank you - how about adding it as:
> This ensures that only the nearest receiver’s members and extensions are accessible within a lambda, preventing unintended interactions between nested scopes.
